eZ/Publish/API/Repository/Tests/RoleServiceTest.php

```diff
@@ -501,6 +518,175 @@ public function testCreateRoleDraftInTransactionWithRollback()
+        }
+
+        return [
+            'without-policies' => [
+                $roleCreateStruct,
+                $roleCopyStruct,
+                [],
+            ],
+            'with-policies' => [
+                $roleCreateStruct,
+                $roleCopyStruct,
+                [$policyCreateStruct1, $policyCreateStruct2],
+            ],
+            'with-limitations' => [
+                $roleCreateStruct,
+                $roleCopyStruct,
+                [$policyCreateStruct1WithLimitations, $policyCreateStruct2],
+            ],
+        ];
+    }
+
+    /**
+     * @dataProvider providerForCopyRoleTests
+     *
+     * @covers \eZ\Publish\API\Repository\RoleService::copyRole
+     * @depends eZ\Publish\API\Repository\Tests\RoleServiceTest::testNewRoleCopyStruct
+     * @depends eZ\Publish\API\Repository\Tests\RoleServiceTest::testLoadRoleByIdentifier
+     *
+     * @throws \eZ\Publish\API\Repository\Exceptions\ForbiddenException
+     * @throws \eZ\Publish\API\Repository\Exceptions\NotFoundException
+     * @throws \eZ\Publish\API\Repository\Exceptions\UnauthorizedException
+     * @throws \eZ\Publish\API\Repository\Exceptions\InvalidArgumentException
+     * @throws \eZ\Publish\API\Repository\Exceptions\LimitationValidationException
+     */
+    public function testCopyRole(RoleCreateStruct $roleCreateStruct, RoleCopyStruct $roleCopyStruct): void
+    {
+        $repository = $this->getRepository();
+
+        $roleService = $repository->getRoleService();
+
+        $roleDraft = $roleService->createRole($roleCreateStruct);
+
+        $roleService->publishRoleDraft($roleDraft);
+        $role = $roleService->loadRole($roleDraft->id);
+
+        $copiedRole = $roleService->copyRole($role, $roleCopyStruct);
+
+        // Now verify that our change was saved
+        $role = $roleService->loadRoleByIdentifier('copiedRole');
+
+        $this->assertEquals($role->id, $copiedRole->id);
+        $this->assertEquals('copiedRole', $role->identifier);
+        $this->assertEmpty($role->getPolicies());
+    }
+
+    /**
+     * Test for the copyRole() method with added policies.
+     *
+     * @dataProvider providerForCopyRoleTests
+     *
+     * @covers \eZ\Publish\API\Repository\RoleService::copyRole
+     * @depends eZ\Publish\API\Repository\Tests\RoleServiceTest::testNewRoleCopyStruct
+     * @depends eZ\Publish\API\Repository\Tests\RoleServiceTest::testLoadRoleByIdentifier
+     *
+     * @throws \eZ\Publish\API\Repository\Exceptions\ForbiddenException
+     * @throws \eZ\Publish\API\Repository\Exceptions\NotFoundException
+     * @throws \eZ\Publish\API\Repository\Exceptions\UnauthorizedException
+     * @throws \eZ\Publish\API\Repository\Exceptions\InvalidArgumentException
+     * @throws \eZ\Publish\API\Repository\Exceptions\LimitationValidationException
+     */
+    public function testCopyRoleWithPolicies(
+        RoleCreateStruct $roleCreateStruct,
+        RoleCopyStruct $roleCopyStruct,
+        array $policies
```

Comment:
For that single one parameter you actually still need PHPDoc, because arrays can contain mixed data (though they shouldn't). AFAICS you can be here even quite specific:
```php
@param \eZ\Publish\API\Repository\Values\User\PolicyCreateStruct[] $policies
```

Reply:
resolved in https://github.com/ezsystems/ezpublish-kernel/pull/3009/commits/f88e2b0caf50b92d986069708a7eef17310a542c